### PR TITLE
Refactor fetching OAuth2 client info, remove cachingActionCreatorFactory

### PR DESCRIPTION
**Pull request by jsnajdr**
Removes the `cachingActionCreatorFactory` abstraction in `state/utils` that was only ever used once and rewrites the `fetchOAuth2ClientData` code without it. The result is a very simple action thunk.

When using WP.com login to login into an OAuth app (i.e., the `client_id` query arg is present), `fetchOAuthClientData` fetches information about the OAuth app so that it can customize the login page. That happens both in SSR (the page.js route handler waits until the fetch is complete, and the results are cached using the `lru` package) and in client-side JS.

**How to test:**
Verify that when loading a SSR-ed login page from URL like this:
```
http://calypso.localhost:3000/log-in?client_id=930&redirect_to=https://wordpress.com/theme?client_id=930
```
then a customized VaultPress login page is returned from server:
Screenshot 2019-10-08 at 12 13 24

Verify that the SSR-ed header for a custom OAuth app (one I registered at developer.wordpress.com) contains the app name:
```
http://calypso.localhost:3000/log-in?client_id=58883&redirect_to=https://wordpress.com/theme?client_id=58883
```
renders as:
Screenshot 2019-10-08 at 12 14 03

(notice that I disabled JS in Chrome Devtools to make sure I see the SSR-ed page)

Verify that the same rendering also happens client-side.

### CODE REVIEW

**Comment by jsnajdr** [comment]
I just rebased this one on top of latest `master`, resolving conflicts with #36470. We both rewrote `createReducer` usage to a normal function 🙂 My rewrite was bigger, as I also changed action types etc.

@sgomes was the usage of `withoutPersistence` in the `state/oauth2-clients/reducer` really necessary? "no persistence" is the default setup, isn't it? I removed it during the rebase, can add it back if it's really needed for some reason I'm missing.

File: client/state/oauth2-clients/reducer.js [hunk 1]
```diff
@@ -52,15 +48,17 @@ export const initialClientsData = {
...
+export default function oauth2Clients( state = initialClientsData, action ) {
 	switch ( action.type ) {
-		case OAUTH2_CLIENT_DATA_REQUEST_SUCCESS: {
-			const { data } = action;
-			const newData = Object.assign( {}, state[ data.id ], data );
-
-			return Object.assign( {}, state, { [ data.id ]: newData } );
-		}
+		case OAUTH2_CLIENT_DATA_RECEIVE:
+			return {
```
Comment (sgomes) [inline client/state/oauth2-clients/reducer.js]: So much cleaner. Well done 👍

File: client/state/oauth2-clients/test/reducer.js [hunk 2]
```diff
@@ -1,66 +1,35 @@
...
 
-	test( 'should throw an error when no parameter is provided', () => {
-		expect( () => reducer() ).to.throw( TypeError );
-	} );
-
-	test( 'should throw an error when no action is provided', () => {
-		expect( () => reducer( {} ) ).to.throw( TypeError );
-	} );
-
-	test( 'should return the current state for an empty action', () => {
```
Comment (sgomes) [inline client/state/oauth2-clients/test/reducer.js]: Any particular reason for removing this and other tests? They don't add a whole bunch, but at least they test that the reducer doesn't behave unexpectedly in edge cases.

File: client/state/oauth2-clients/test/reducer.js [hunk 3]
```diff
@@ -111,20 +80,4 @@ describe( 'reducer', () => {
 			},
 		} );
 	} );
-
-	test( 'should not persist state', () => {
```
Comment (sgomes) [inline client/state/oauth2-clients/test/reducer.js]: This seems useful to ensure. Perhaps testing it with a solo `combineReducers`?

**Review by sgomes — APPROVED** [review]
The code generally looks good, but I'm concerned we're removing so many tests. Would it be possible to rewrite them instead? I'll leave it as your call, though.

**Comment by sgomes** [comment]
> @sgomes was the usage of `withoutPersistence` in the `state/oauth2-clients/reducer` really necessary? "no persistence" is the default setup, isn't it? I removed it during the rebase, can add it back if it's really needed for some reason I'm missing.

It's likely unnecessary, but as a codemod I wanted to follow the incumbent behaviour as closely as possible, and `createReducer` did add a `withoutPersistence` wrapper. I think it's fine to remove it as we go, now, with the appropriate test changes 👍

File: client/state/oauth2-clients/test/reducer.js (hunk 2 above)
Comment (jsnajdr) [reply client/state/oauth2-clients/test/reducer.js]: The tests that test if the reducer throws error if called with no or invalid parameters: I think they don't do anything useful. Redux never calls them that way, so it doesn't matter at all whether they throw or not.

The "return current state for an empty action" test is testing the same trivial thing as the "return current state for an unknown action type" -- namely the `default: return state` line. That's what every reducer that's not completely broken will do.

Handling `SERIALIZE` and `DESERIALIZE` actions is no longer a job of this reducer. It was when it was implemented with `createReducer`, but now the persistence wrapper will be added by `combineReducer` by default. See [here](https://github.com/Automattic/wp-calypso/blob/8cf852867340c735948e9dfec78ebce34cf4cab7/client/state/utils/reducer-utils.js#L136) how the `setupReducerPersistence` wraps every plain reducer with `withoutPersistence` before combining.

Therefore, `SERIALIZE` and `DESERIALIZE` is fully covered by the `state/utils` tests.

File: client/state/oauth2-clients/test/reducer.js (hunk 3 above)
Comment (jsnajdr) [reply client/state/oauth2-clients/test/reducer.js]: That would just duplicate the unit tests of `combineReducers`.

**Pull request merged**